crypto/des/set_key.c

```diff
@@ -62,15 +64,18 @@ void DES_set_odd_parity(DES_cblock *key)
         (*key)[i] = odd_parity[(*key)[i]];
 }
 
+/*
+ * Check that a key has the correct parity.
+ * Return 1 if parity is okay and 0 if not.
+ */
 int DES_check_key_parity(const_DES_cblock *key)
 {
     unsigned int i;
+    unsigned char res = 0377;
 
-    for (i = 0; i < DES_KEY_SZ; i++) {
-        if ((*key)[i] != odd_parity[(*key)[i]])
-            return 0;
-    }
-    return 1;
+    for (i = 0; i < DES_KEY_SZ; i++)
+        res &= constant_time_eq_8((*key)[i], odd_parity[(*key)[i]]);
```

Comment:
accessing the odd_parity array with a secret key byte is actually not constant time

Reply:
its probably overkill but you could do something like that in constant time:
```C
  unsigned b = (*key)[i];
  b ^= b >> 4;
  b ^= b >> 2;
  b ^= b >> 1;
  res &= constant_time_eq(b & 1, 1);
```

Reply:
Personally I prefer this over the table look up approach

Reply:
Okay, I'll look at changing this tomorrow.  @bernd-edlinger, I was aware that the parity array look up wouldn't be constant time.  I expected it to be far less gratuitous than what was present.

BTW, gcc and presumably other compilers have a `__builtin_parity` intrinsic.

When are compilers going to add a _constant_time_ attribute.  I expect the [movcc](https://github.com/xoreaxeaxeax/movfuscator) would come close for all inputs.

Reply:
Updated to use the above code instead of a table lookup.

Reply:
I think it should be possible to compute all of the parities in parallel.  I doubt it's worth the effort for DES though.